### PR TITLE
Reintroducing caching

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR reimplements caching in chemprop. It was originally _not_ reimplemented because non-cached v2 is faster than cached v1. However, that doesn't mean that caching is now pointless. In fact, it's even more potent now because the `MolGraph` object is smaller so you can cache larger datasets. In my experience, caching a dataset can get me up to ~120 batches/s using `batch_size=128`

## Changes
- uses `Generic`s to parameterize both `{Molecule,Rxn}MolGraphFeaturizers`. There's more work to be done moving the whole package over to utilize generics (and saving a lot of code), but that's out of the scope of this PR. The small changes here were necessary to properly add generic type support to the caching class.
- adds a new `MolGraphCacheFacade` class. The name is meant to imply that the base class simply provides the "façade" of a cache of `MolGraph`s, **not** whether it actually implements one. The cache is implemented using a sequence-like interface given that the client objects are the `{Molecule,Rxn}MolGraphDataset` class's `__getitem__` calls. That is, calling `cache[i]` get's the `i`-th item from the cache.
- adds a new `cache` property to the `{Molecule,Rxn}MolGraphDataset` classes. This is set to `False` in the `__init__`. Setting it to `True` then caches the entire dataset. The implication of this means that a user must set `cache=True` **after** initializing their dataset to actually cache it.
- adds a new mock-based test to ensure that the underlying `mg_cache` is being used

**"Why is `cache=False` not configurable in `__init__`?"**
The idea behind this design is that if a user wants to cache their dataset, they might set `cache=True`, caching all their `MolGraph`s, and then at a later point scale their additional atom/bond features. This would require recomputing the cache, but it probably won't be clear to users that they would need to reinitialize the underlying cache.

## Example Usage
To enable the cache, a user must set `cache=True`, like so:
```python
dset = MoleculeDataset(...)
# do some descriptor scaling here
dset.cache = True
```

## Questions
I'm ambivalent as to whether we expose `cache` via the `__init__` function, but for right now this is good to go.

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by davidegraff** [comment]
tests are failing but they're not related to any of the changes I've made somehow...

**Comment by kevingreenman** [comment]
thanks for the PR! It looks like the tests that are failing are Mac OS (has been failing for a while due to the torch-scatter installation issue) and Python 3.12, which we just started testing on in #695 and do not officially support yet. So I don't think there's anything to worry about regarding the tests for this PR.

File: chemprop/featurizers/molgraph/base.py
Comment (davidegraff) [inline chemprop/featurizers/molgraph/base.py:28]:
```suggestion
        input : T
```

File: chemprop/featurizers/molgraph/cache.py
Comment (davidegraff) [inline chemprop/featurizers/molgraph/cache.py:8]:
```suggestion
from chemprop.featurizers.molgraph.base import MolGraphFeaturizer, T
```

File: chemprop/featurizers/molgraph/cache.py
Comment (davidegraff) [inline chemprop/featurizers/molgraph/cache.py:11]:
```suggestion
```

**Review by hwpang — COMMENTED** [review]
Thanks for the PR! I'm fine with requiring the user to set cache after the dataset is initialized. Could you add the use of `cache` somewhere into the documentation and note that `cache` must be set after the inputs/targets normalization?

File: chemprop/featurizers/molgraph/base.py
Comment (davidegraff) [reply chemprop/featurizers/molgraph/base.py:28]: `T` is an abstract type and is a common notation in programming. There's no need to actually constrain it as I defined it above. IMO it should not be documented.

**Review by JacksonBurns — DISMISSED** [review]
Before I review the code itself, could you add a description somewhere in the v2 docs of what this does? Even just a couple paragraphs describing at a high level what it does, when it is useful, and how it is exposed to the user would be good. I never found the [v1 caching docs](https://chemprop.readthedocs.io/en/latest/tutorial.html#caching) especially helpful...

**Comment by hwpang** [comment]
If the user doesn't intend to use `cache` but scales the inputs/ targets, the current implement would require the users to set `dset.cache=False` in order for the cached data to be updated, is that correct?

**Comment by davidegraff** [comment]
@JacksonBurns- I don't really see a good spot to place it right now. The python tutorial of the CLI is seemingly deprecated for some reason but I would ideally put this there. In the meantime, I've added some additional docs to both classes.

**Comment by davidegraff** [comment]
> If the user doesn't intend to use `cache` but scales the inputs/ targets, the current implement would require the users to set `dset.cache=False` in order for the cached data to be updated, is that correct?

good catch! I didn't think of that, but 6e737f7 should fix that now.

**Review by hwpang — APPROVED** [review]
I'm happy with the current changes. I would like this to be added into the documentation, but from the comment above it seems that currently there is not an appropriate place to add the documentation yet. We can merge this in and open an issue to keep track of this.

**Comment by davidegraff** [comment]
> > @JacksonBurns I don't really see a good spot to place it right now.
> 
> I see this as an **especially** compelling reason to add the documentation (create a place to put it and then put it there), but I won't stand in the way of progress.

I agree, but it seems out-of-scope for this PR to "undeprecate" the Python module docs _and_ add a new feature. I wouldn't disagree with creating an issue to do that as a blocker to this PR. As far as the CLI, this feature is not exposed via the CLI, so it doesn't feel correct to add it to those docs either.

File: tests/unit/featurizers/test_molgraph.py
Comment (KnathanM) [inline tests/unit/featurizers/test_molgraph.py:10]: Probably fine to import this explicitly, but then why not also do `from chemprop.featurizers.molgraph.molecule import SimpleMoleculeMolGraphFeaturizer` explicitly. I think it would be better to group `MolGraphFeaturizer` with the other imports.

**Review by KnathanM — COMMENTED** [review]
### Checking my understanding
When a dataset is created, cache is set to False which calls the `cache.setter` and sets the `self.mg_cache` to be an object that will calculate the molgraphs on the fly. Then when the V_fs and E_fs are normalized, `self.mg_cache` will be updated to use the new scaled features when calculating the molgraphs. At this point, if the users set `dataset.cache = False` an identical object to calculate the molgraph is recreated, but if they set `dataset.cache = True`, `self.mg_cache` will be overwritten with an object that has all the molgraphs computed and saved to memory. Setting `dataset.cache = True` again will recalculate all the molgraphs to overwrite `self.mg_cache`, which you shouldn't do if you don't actually need to update anything in the cache. 

But if you set `dataset.cache = True` before normalizing V_fs and E_fs, then when you do normalize them your cache will be updated. Thus it is highly advisable to normalize inputs before caching to avoid the extra work of caching needlessly. I think that should be clearly stated in the documentation. I will think about where this would go in the documentation, per the other comments on this PR. 

### Naming
I'm a bit confused by `MolGraphCacheOnTheFly`'s name. To me calculating the features on the fly is the opposite of caching (calculating them all once at the beginning, storing them, and then looking them up as needed). So maybe a better name would be `MolGraphOnTheFly`. 
I also think using `self.mg_cache` in `datasets.py` is a bit misleading for the same reason that the molgraphs could be in a cache or could be computed as you go. Maybe calling it `self.mgs` instead is more clear. 

### Possible alternative
What do you think about having more distinction between caching molgraphs and computing on the fly? We could rename `_init_cache` to `update_mgs` like 
```
def update_mgs(): 
    self.mgs = (MolGraphCache if self.cache else MolGraphOnTheFly)(self.mols, ...)
```
Also separate the caching function of `self.cache` from its function of storing whether to cache by introducing `set_cache()`.
```
@property
def cache(self) -> bool:
    return self.__cache

@cache.setter
def cache(self, cache: bool):
    self.__cache = cache
	
def set_cache():
    self.cache = True
    update_mgs()
```
When a user wants to calculate and cache the molgraphs, then they would call `dataset.set_cache()`. I'm not tied to the name `set_cache` but to me running something like `dataset.set_cache()` multiple times more clearly will reset the cache each time vs running `dataset.cache = True` multiple times. 
You could also add an opposite function to set the molgraphs back to on the fly calculation. It feels nicely symmetric, but not sure if that would be needed.
```
def clear_cache():
    self.cache = False
    update_mgs()
```

**Comment by davidegraff** [comment]
> ### Checking my understanding
> When a dataset is created, cache is set to False which calls the `cache.setter` and sets the `self.mg_cache` to be an object that will calculate the molgraphs on the fly. Then when the V_fs and E_fs are normalized, `self.mg_cache` will be updated to use the new scaled features when calculating the molgraphs. At this point, if the users set `dataset.cache = False` an identical object to calculate the molgraph is recreated, but if they set `dataset.cache = True`, `self.mg_cache` will be overwritten with an object that has all the molgraphs computed and saved to memory. Setting `dataset.cache = True` again will recalculate all the molgraphs to overwrite `self.mg_cache`, which you shouldn't do if you don't actually need to update anything in the cache.

Correct. Everytime you set `dataset.cache = True`, you will "recalculate" the cache. But due to the object-oriented nature here, this could effectively be a no-op (if you're using an on-the-fly "cache").

> But if you set `dataset.cache = True` before normalizing V_fs and E_fs, then when you do normalize them your cache will be updated. Thus it is highly advisable to normalize inputs before caching to avoid the extra work of caching needlessly. I think that should be clearly stated in the documentation. I will think about where this would go in the documentation, per the other comments on this PR.

👍 

> ### Naming
> I'm a bit confused by `MolGraphCacheOnTheFly`'s name. To me calculating the features on the fly is the opposite of caching (calculating them all once at the beginning, storing them, and then looking them up as needed). So maybe a better name would be `MolGraphOnTheFly`. I also think using `self.mg_cache` in `datasets.py` is a bit misleading for the same reason that the molgraphs could be in a cache or could be computed as you go. Maybe calling it `self.mgs` instead is more clear.

This where polymorphism and object-oriented design comes into play. For the purposes of the `*Dataset` classes, they don't actually _know_ what the underlying cache implementation is. Rather, all they see is something _that looks like and acts like a cache_. This is an important detail to hide to keep the interaction between these two classes a high level. This is why the both the `MolGraphCache` and `MolGraphCacheOnTheFly` derive from the `MolGraphCacheFacade` class. The top-level class **does not guarantee** that it provides a true cache implementation--just that it provides something that provides the façade of a cache. I disagree with with both of the suggested name changes for this reason:

- `MolGraphOnTheFly` implies that that the class implements a `MolGraph`. It does not. Instead, it is closest to supplying a `list[MolGraph]`.
- `self.mgs` implies that this attribute is `list[MolGraph]`. Again, this is not logically correct because the attribute only provides access to a façade of the cache. Naming it like this suggests it provides something that it doesn't.

> ### Possible alternative
> What do you think about having more distinction between caching molgraphs and computing on the fly? We could rename `_init_cache` to `update_mgs` like
> 
> ```
> def update_mgs(): 
>     self.mgs = (MolGraphCache if self.cache else MolGraphOnTheFly)(self.mols, ...)
> ```
> 
> Also separate the caching function of `self.cache` from its function of storing whether to cache by introducing `set_cache()`.
> 
> ```
> @property
> def cache(self) -> bool:
>     return self.__cache
> 
> @cache.setter
> def cache(self, cache: bool):
>     self.__cache = cache
> 	
> def set_cache():
>     self.cache = True
>     update_mgs()
> ```
> 
> When a user wants to calculate and cache the molgraphs, then they would call `dataset.set_cache()`. I'm not tied to the name `set_cache` but to me running something like `dataset.set_cache()` multiple times more clearly will reset the cache each time vs running `dataset.cache = True` multiple times. You could also add an opposite function to set the molgraphs back to on the fly calculation. It feels nicely symmetric, but not sure if that would be needed.
> 
> ```
> def clear_cache():
>     self.cache = False
>     update_mgs()
> ```

The suggested change to the `cache` property is pointless. Why would we need to make `cache` a "private" variable? The point of properties is generally to establish some invariance in the object's behavior. Here, the purpose is that when a user sets `cache = True`, it actually starts using a true cache. With your change, the two can be decoupled. That is, a user can set the attribute to `True` but not actually be using a true cache, thus decoupling the attribute from its intended use. I'm not opposed to changing the idiom here if you feel very strongly about it. IMO, setting `cache = True` and actually setting the cache seems idiomatic to me, but I can swap it out for a method: `init_cache()`, `initialize_cache()`, `enable_cache()`, etc. `set_cache()` is a poor name IMO because it sounds similar to the getter/setter-paradigm, which your comment suggests moving away from.

File: chemprop/featurizers/molgraph/base.py
Comment (davidegraff) [reply chemprop/featurizers/molgraph/base.py:20]: This is a non-issue. There's **zero** reason why one would want to use the native `input()` function _inside_ module-level code. Moreover, `input` is the standard argument name for `torch.nn.Module.forward()`. If it's good enough for them, then it's certainly good enough for us.

File: chemprop/featurizers/molgraph/base.py
Comment (davidegraff) [reply chemprop/featurizers/molgraph/base.py:28]: `SimpleMoleculeMolGraphFeaturizer.__call__` will inherit this docstring. It seems like a lot of wasted documentation just to change a single type. The reason `CondensedGraphOfReactionFeaturizer` overwrite this docstring is because we need to note that the additional arguments are unused.

**Review by KnathanM — APPROVED** [review]
After more thought, I agree that your way of setting and resetting the cache is idiomatic. I still think "cache on the fly" is an oxymoron, but oxymorons do have their place in communication ("old news", "loyal opposition", etc.). So I am happy to merge these changes in. @JacksonBurns I see you have changes requested, is that still true?

In terms of documentation, I think it could go in the training.ipynb example notebook in a cell after the datasets are created and scaled. Does this description sound corret?

By default, the molecules are featurized to MolGraph objects as they are needed. If the dataset is not too large, it is faster to create all the MolGraph objects once at the beginning and cache them to memory. This can be done using `dataset.cache = True`. Note that if the the additional atom or bond features are set or scaled after caching, the cache will automatically be recalculated. Thus it is recommended to cache _after_ normalizing.
```
train_dset.cache = True
val_dset.cache = True
test_dset.cache = True
```

If that sounds good to you David, feel free to add it to the PR. I'm also happy to open another PR myself and add it.

**Comment by davidegraff** [comment]
Yeah I agree it's an oxymoron, but that's sort of the point: behave like it _is_ a cache but to be clear that it is _not_ a cache by way of implementation. IMO actually adding this feature to the CLI and examples would best be a separate, small PR. If we squash commit then we can clearly see in the commit history the small changes of when we (i) _added_ this feature and (ii) _started using_ this feature.

**Comment by JacksonBurns** [comment]
> @JacksonBurns I see you have changes requested, is that still true?

I'll defer to you and @hwpang on this one. I would (somewhat selfishly) like to merge the CI overhaul PR first and check that this passes on all platforms before merging

**Comment by KnathanM** [comment]
@davidegraff Can you check if maintainers are allowed to push to your branch? We have merged in the CI changes and there are two small merge conflicts that I have resolved on my local that I'd like to push to update your branch.

**Comment by davidegraff** [comment]
feel free to make edits

**Pull request merged**
